jwst/pipeline/calwebb_image3.py

```diff
@@ -41,77 +40,63 @@ class Image3Pipeline(Pipeline):
                  }
 
-    def process(self, input):
+    def process(self, input_data):
         """
         Run the Image3Pipeline
 
         Parameters
         ----------
-        input: Level3 Association, or ModelContainer
+        input_data: Level3 Association, or ModelContainer
             The exposures to process
         """
-
         self.log.info('Starting calwebb_image3 ...')
 
         # Only load science and background members from input ASN
         asn_exptypes = ['science', 'background']
-        input_models = datamodels.open(input, asn_exptypes=asn_exptypes)
-
-        # If input is an association, set the output to the product name.
-        try:
-            self.output_file = input_models.meta.asn_table.products[0].name
-        except AttributeError:
-            pass
-
-        # Check if input is single or multiple exposures
-        is_container = isinstance(input_models, datamodels.ModelContainer)
-        try:
-            has_groups = len(input_models.group_names) > 1
-        except Exception:
-            has_groups = False
-        if is_container and has_groups:
-            if is_moving_target(input_models):
-                self.log.info("Assigning WCS to a Moving Target exposure.")
-                input_models = self.assign_mtwcs(input_models)
-            else:
-                self.log.info("Aligning input images...")
-                input_models = self.tweakreg(input_models)
-
-            self.log.info("Matching sky values across all input images...")
-            input_models = self.skymatch(input_models)
-
-            self.log.info("Performing outlier detection on input images...")
-            input_models = self.outlier_detection(input_models)
-
-            if input_models[0].meta.cal_step.outlier_detection == 'COMPLETE':
-                self.log.info("Writing Level 2c images with updated DQ arrays...")
-                # Set up Level 2c suffix to be used later
-                asn_id = input_models.meta.asn_table.asn_id
-                suffix_2c = '{}_{}'.format(asn_id, 'crf')
-                for model in input_models:
-                    self.save_model(
-                        model,
-                        output_file=model.meta.filename,
-                        suffix=suffix_2c
-                    )
-
-        self.log.info("Resampling images to final result...")
-        result = self.resample(input_models)
-
-        try:
-            result.meta.asn.pool_name = input_models.meta.asn_table.asn_pool
-            result.meta.asn.table_name = os.path.basename(input)
-            result.meta.filename = input_models.meta.asn_table.products[0].name
-        except Exception:
-            pass
-
-        self.save_model(result, suffix=self.suffix)
-
-        self.log.info("Creating source catalog...")
+
+        # Configure settings for saving results files
+        self.outlier_detection.suffix = 'crf'
+        self.outlier_detection.save_results = self.save_results
```

Comment:
This looks like it's assigning the value of `save_results` that applies to the whole pipeline to the individual steps. What happens if a user specifies an override, e.g. `--steps.outlier_detection.save_results=False`? Is the user's value going to be overridden by the pipeline setting?

Reply:
I think it would, yes, and is probably not desired now that I think of it. However, this is the same practice used in the `Spec3Pipeline` too. 

To fix this then, I suppose that I would add the suffix back into the spec string, and remove those explicit assignments.

Reply:
Then, I guess that we'd also need to fix that in `Spec3Pipeline` too

Reply:
Someone could/should run a quick test of seeing if the command-line overrides actually work.

Reply:
I think that the intent here was that the pipeline should produce these output files by default, but now that I'm thinking about it, I'm not actually sure how you'd control that at the pipeline level while still allowing for overrides for individual steps. 

If you removed those assignments in `process`, the only way to set it (I think.. isn't the default for `save_results` for the steps `False`?) would be to configure it after instantiation or with a config file, right?

Another approach that I see in other pipelines is to add additional arguments in `spec` for use in switching saving results on/off

Reply:
The default setting for `save_results` in the `stpipe` `step` class is `False`. And this same operation of setting the value for individual steps, like `resample`, to the step value is also used in some other pipelines, like `calwebb_image2`. So now I don't understand how those pipeline are actually managing to create output products from these individual steps when the default seems to be `False`.  @stscieisenhamer any wisdom here?

Reply:
`save_results` defaults to `False` for all steps/pipelines.  When you use `strun` on the command line, it sets `save_results` for the pipeline (or whatever strun is running) to `True`.  So I think the above is valid.  It allows for nothing to be written out if being run from within python, and for the normal things to be written out when running using `strun`.

That said, if you want the user to be able to specify turning off writing out the `outlier_detection` results, then you could check first if the user has set it?  But that's tricky because it's an order-of-operations issue.  Hmm.

Reply:
What if we added an argument in `spec`, something like: 
```python
spec = """
    default_save_results = bool(default=True)
"""
```

Combined with: 

```python
if self.default_save_results:
    # Configure settings for saving results files
    self.outlier_detection.suffix = 'crf'
    self.outlier_detection.save_results = self.save_results

    self.resample.suffix = 'i2d'
    self.resample.save_results = self.save_results

    self.source_catalog.save_results = self.save_results
```

That way, users could have the same "default results" output just like when executing from the command line, but they could also disable that and set things manually or with a config file. Seems like that might cover both cases, while by default providing the output products that the stakeholders seem to expect

Reply:
> `save_results` defaults to `False` for all steps/pipelines. When you use `strun` on the command line, it sets `save_results` for the pipeline (or whatever strun is running) to `True`. So I think the above is valid. It allows for nothing to be written out if being run from within python, and for the normal things to be written out when running using `strun`.
> 
> That said, if you want the user to be able to specify turning off writing out the `outlier_detection` results, then you could check first if the user has set it? But that's tricky because it's an order-of-operations issue. Hmm.

Ah, that's right. The default is `False` for the purpose of running interactively within python, where you normally just want the results passed back in a data model and not saved to disk. When running from command line it's automatically reset to `True`, otherwise there's no way to get the results.

The default of `True` is necessary for all the steps that create required products, which includes `resample`, `outlier_detection`, `source_catalog`, `extract_1d`, etc. (the Ops environment always wants those results). I suppose a user might want to be able to turn off some of those occasionally, but it's more import to allow them to simply turn on the saving of results for intermediate steps that don't normally save their products.

Reply:
Just to provide resolution to this thread: 

This is actually a known issue ( #2824 #2723 ), and is outside of the scope of this PR, and will be addressed later.